src/lib/storage/table.cpp

```diff
@@ -56,30 +60,77 @@ uint16_t Table::column_count() const { return _column_names.size(); }
-  for (ChunkID chunk_index = (ChunkID)0; chunk_index < _chunks.size(); chunk_index++) {
+  for (auto chunk_index = ChunkID{0}; chunk_index < _chunks.size(); chunk_index++) {
     row_count += _chunks[chunk_index]->size();
   }
   return row_count;
 }
 
-ChunkID Table::chunk_count() const { return (ChunkID)_chunks.size(); }
+ChunkID Table::chunk_count() const { return ChunkID{_chunks.size()}; }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
   auto iterator = std::find(_column_names.begin(), _column_names.end(), column_name);
-  if (iterator != _column_names.end()) return (ColumnID)std::distance(_column_names.begin(), iterator);
+  if (iterator != _column_names.end()) return ColumnID{std::distance(_column_names.begin(), iterator)};
   throw std::exception();
 }
 
 uint32_t Table::max_chunk_size() const { return _maximum_chunk_size; }
 
 const std::vector<std::string>& Table::column_names() const { return _column_names; }
 
-const std::string& Table::column_name(ColumnID column_id) const { return _column_names[column_id]; }
+const std::string& Table::column_name(ColumnID column_id) const {
+  DebugAssert(column_id < _column_names.size(), "There exists no column with the given ID.");
+  return _column_names[column_id];
+}
 
-const std::string& Table::column_type(ColumnID column_id) const { return _column_types[column_id]; }
+const std::string& Table::column_type(ColumnID column_id) const {
+  DebugAssert(column_id < _column_names.size(), "There exists no column with the given ID.");
+  return _column_types[column_id];
+}
 
 Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks[chunk_id]; }
 
 const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks[chunk_id]; }
 
+std::mutex chunk_access_mutex;
+
+void Table::compress_chunk(ChunkID chunk_id) {
+  auto new_chunk = std::make_shared<Chunk>();
+  auto old_chunk = _chunks.at(chunk_id);
+
+  std::vector<std::thread> threads(old_chunk->column_count());
+  std::vector<std::shared_ptr<BaseSegment>> compressed_segments(old_chunk->column_count());
+
+  // Lambda function which compresses a given uncompressed segment (with a given type) and writes the output to the
+  // given position in the compressed_segments vector
+  auto compress_segment = [&compressed_segments](const std::string& column_type,
+                                                 const std::shared_ptr<BaseSegment>& uncompressed_segment,
+                                                 const ColumnID& segment_index) {
+    compressed_segments[segment_index] =
+        make_shared_by_data_type<BaseSegment, DictionarySegment>(column_type, uncompressed_segment);
+  };
+
+  // Start one thread for each of the segments in order to compress it
+  for (auto segment_index = ColumnID{0}; segment_index < old_chunk->column_count(); segment_index++) {
+    threads[segment_index] =
+        std::thread(compress_segment, column_type(segment_index), old_chunk->get_segment(segment_index), segment_index);
```

Comment:
I find this slightly confusing as you pass arguments both using the lambda capture and the thread arguments. Personally, I would have used an inline lambda (`std::thread([&, segment_index](){})`), but that's also a question of personal preference.